### PR TITLE
std: add math.Interval

### PR DESCRIPTION
**Pull request by nektro**
Implements https://en.wikipedia.org/wiki/Interval_(mathematics) for Zig.

`T` may be any of `uN`, `iN`, `f16`, `f32`, `f64`, `f128`.

Properly handles all range types:
- `[a, b]`
- `[a, b)`
- `(a, b]`
- `(a, b)`
- `a` is `-inf`
- `b` is `+inf`

### CODE REVIEW

**Comment by mattbork** [comment]
Maybe `open` should be ignored when an endpoint is at infinity. For example, if `T` is an integer type and `from` and `to` are both null then `std.math.maxInt(T)` will only be inside the interval if `to` is also closed, whereas I'd expect the finite `maxInt(T)` to be inside both `[-∞,∞]` and `[-∞,∞)`. The `minInt` isn't affected because of the order of inf cases.

On the other hand, for floats, `std.math.inf(T)` probably should only be in `[-∞,∞]` and not in `[-∞,∞)`. Or perhaps floating inf should be always be considered outside any interval.

File: lib/std/math/Interval.zig [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+            const from = self.from.n orelse min;
+            const to = self.to.n orelse max;
+            const left = if (self.from.open) x > from else x >= from;
+            const right = if (self.to.open) x < to else x <= to;
+            return left and right;
+        }
+    };
+}
+
+//
```
Comment (g-w1) [inline lib/std/math/Interval.zig:61]: Not sure why this comment is hanging out here?

**Comment by nektro** [comment]
@mattbork -inf and +inf are handled specially here

https://github.com/ziglang/zig/pull/8745/files#diff-129101bf3397b523d6149596c15ed27e9abadbb8a2e436923b145edda9be83c4R40-R51

File: lib/std/math/Interval.zig (hunk 1 above)
Comment (nektro) [reply lib/std/math/Interval.zig:61]: I added this when I was working on it to separate the public API/implementation from the tests. If this is against style, I can remove it.

**Comment by mattbork** [comment]
@nektro I know, but if T is an integer, and `from` and `to` are *both* null, then when the special case for `(-inf` is run:

```
const to = self.to.n orelse max;
const right = if (self.to.open) x < to else x <= to;
return right;
```

the local `to` will be set to max. So if `self.to.open` is true, then the finite `std.math.maxInt(T)` will be excluded from the interval. As I said, I would expect any finite integer to be conatined in both [-∞,∞] and [-∞,∞). However,

```
var x = Interval(u32){
    .from = .{ .n = null, .open = false  },
    .to =   .{ .n = null, .open = true },
};

try std.testing.expect(x.contains(std.math.minInt(u32))); // Ok
try std.testing.expect(x.contains(std.math.maxInt(u32))); // Fails
```

**Comment by nektro** [comment]
Ah, good catch

File: lib/std/math/Interval.zig [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        from: Endpoint,
+        /// The right hand side of the Interval.
+        /// n as null here represent infinity.
+        /// open determines if the interval should be open (exclusive) or closed (inclusive). 
+        to: Endpoint,
+
+        pub const Endpoint = struct {
+            n: ?T,
+            open: bool,
+        };
```
Comment (marler8997) [inline lib/std/math/Interval.zig:20]: It seems to me that the optional modifier should be put on the fields `from` and `to` rather than `n`, since I'm not sure that the "open" field means anything when `n` is null.

Or you could use a tagged union like this:

```zig
const Endpoint = union(enum) {
    Infinity: void,
    Bounded: struct { val: T, open: bool },
};
```

This is another variation that could be used.
```zig
const Endpoint = union(enum) {
    Infinity: void,
    Open: T,
    Closed: T,
};
```

Here's what it initializing a field would look like using the various representations:
```zig
// current
const unbounded = Interval(u8) { .from = .{ .n = null, .open = undefined }, .to = .{ .n = null, .open = undefined } };
// optional field
const  unbounded = Interval(u8) { .from = null, .to = null };
// tagged union
const unbounded = Interval(u8) { .from = EndPoint.Infinity, .to = EndPoint.Infinity };
```

Comment (nektro) [reply lib/std/math/Interval.zig:20]: I like this a lot, I'll update

Comment (nektro) [reply lib/std/math/Interval.zig]: updated 👍

File: lib/std/math/Interval.zig [hunk 3]
```diff
@@ -0,0 +1,110 @@
+const std = @import("../std.zig");
+
+/// Create an interval of type T.
+/// See https://en.wikipedia.org/wiki/Interval_(mathematics) for more info.
+pub fn Interval(comptime T: type) type {
```
Comment (daurnimator) [inline lib/std/math/Interval.zig:5]: What if open/closed-ness was a comptime property of the type?

File: lib/std/math/Interval.zig [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        /// n as null here represents infinity.
+        /// open determines if the interval should be open (exclusive) or closed (inclusive). 
+        from: Endpoint,
+        /// The right hand side of the Interval.
+        /// n as null here represent infinity.
+        /// open determines if the interval should be open (exclusive) or closed (inclusive). 
+        to: Endpoint,
+
+        pub const Endpoint = union(enum) {
+            Infinity,
```
Comment (daurnimator) [inline lib/std/math/Interval.zig:18]: this doesn't make sense for floating point types

File: lib/std/math/Interval.zig [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+            Closed: T,
+            Open: T,
+
+            pub fn n(self: Endpoint) ?T {
```
Comment (daurnimator) [inline lib/std/math/Interval.zig:22]: why `n`? (leave a doc-comment?)

File: lib/std/math/Interval.zig [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+            const right = if (self.to == .Open) x < to else x <= to;
+            return left and right;
+        }
+    };
+}
+
+fn runTest(comptime T: type, from: Interval(T).Endpoint, to: Interval(T).Endpoint) !void {
+    var x = Interval(T){ .from = from, .to = to };
+
+    try std.testing.expect(x.contains(0) == (x.from == .Infinity));
```
Comment (daurnimator) [inline lib/std/math/Interval.zig:77]: use `expectEqual`

**Comment by MasterQ32** [comment]
Can you please add tests for certain corner cases with integer types? Using `0…255` inclusive for `u8`, `-128…127` inclusive `i8`, `0…0` for `u0`, `0…1` for `u1` and `-1…0` for `i1`. These ranges make sure you handle your corner cases correctly and do not accidently widen or overflow a interval

File: lib/std/math/Interval.zig (hunk 4 above)
Comment (LemonBoy) [reply lib/std/math/Interval.zig:18]: Unless you're working with bigints it doesn't even make sense for non-fp types, the number of bits is finite and the upper/lower limit is well known.

**Comment by LemonBoy** [comment]
Interesting idea, here's a few notes:
* The `lower <= upper` precondition is never enforced, you either need to litter every method with assert calls or provide some `init` method (and make it fallible, that's also ugly)
* Many useful methods are missing (eg. arithmetic ops, a check for interval-contains-interval, an `empty` method, equality (via `eql`) and comparison (using `math.Order`), intersection)

Check out this [nice design document](https://www-sop.inria.fr/members/Sylvain.Pion/cxx/interval/interval.pdf), it's IMO pretty complete.

On the other hand, what's the use case for interval arithmetic and why does it deserve a place in the stdlib? IOW what's the elevator pitch?

**Comment by nektro** [comment]
@MasterQ32 good idea, I'll add those

@LemonBoy thanks for the reference :) the elevator pitch would be that it works beyond arithmetic. my initial driver to have this was to have a stdlib construct for Ranges and future work on this would add support for non-number types and support for (maybe in this or in another type) negative ranges

**Comment by nektro** [comment]
for instance, maybe instead of asserting that `T` is a number, it could check if `T` is a number or `T` has `.compareTo`, `.posInf`, and `negInf`

File: lib/std/math/Interval.zig (hunk 5 above)
Comment (nektro) [reply lib/std/math/Interval.zig:22]: good catch, this isn't super self explanatory unless you look at the code. it's a helper to get the `T` value out of `.Open` or `.Closed` if those are set or null if its `.Infinity`

File: lib/std/math/Interval.zig (hunk 6 above)
Comment (nektro) [reply lib/std/math/Interval.zig:77]: okay 👍

File: lib/std/math/Interval.zig (hunk 3 above)
Comment (nektro) [reply lib/std/math/Interval.zig:5]: because its defined in the `Endpoint` struct and its not something that inherently needs to be an invariant so I left it there in the struct over having an `.init()` function

File: lib/std/math/Interval.zig (hunk 4 above)
Comment (nektro) [reply lib/std/math/Interval.zig:18]: why doesn't it make sense?

**Comment by LemonBoy** [comment]
> so that I could represent codepoint ranges for unicode blocks in a library I'm working on

Side note, look into folded tries for that purpose, it's quite space/time efficient as representation and is currently employed by many (all?) unicode libraries.

File: lib/std/math/Interval.zig (hunk 4 above)
Comment (xackus) [reply lib/std/math/Interval.zig:19]: You can just make a closed endpoint with infinity (fp) or the max/min possible value (int).

Comment (nektro) [reply lib/std/math/Interval.zig]: perhaps but including `.infinity` makes intent clearer when constructing the value and removes a branch when in `.contains(x)`

Comment (xackus) [reply lib/std/math/Interval.zig]: Just make a function that returns the infinity value.
I don't see how it would result in less branches. LLVM is definitely smart enough to optimize out `num <= math.maxInt(T)`.
In the general case there are more branches, because you have to test for `.infinity`.

Comment (nektro) [reply lib/std/math/Interval.zig]: thanks for the insight everyone, contains has now been cleaned up significantly

**Comment by andrewrk** [comment]
Thanks for the submission. This looks like a good candidate for a third party package to me - if we want to continue the discussion of adding it to the std lib, let's do that in a proposal issue.
